File: grpc/clients/traffic_policy.go
// Copyright 2019 Smart-Edge.com, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package clients

import (
	"fmt"

	cce "github.com/smartedgemec/controller-ce"
	elapb "github.com/smartedgemec/controller-ce/pb/ela"
)

func toPBTrafficPolicy(id string, tp *cce.TrafficPolicy) *elapb.TrafficPolicy {
	pbPolicy := &elapb.TrafficPolicy{
		Id: id,
	}

	if tp != nil {
		for _, rule := range tp.Rules {
			pbPolicy.TrafficRules = append(
				pbPolicy.TrafficRules, toPBTrafficRule(rule))
		}
	}

	return pbPolicy
}

func toPBTrafficRule(tr *cce.TrafficRule) *elapb.TrafficRule {
	return &elapb.TrafficRule{
		Description: tr.Description,
		Priority:    uint32(tr.Priority),
		Source:      toPBTrafficSelector(tr.Source),
		Destination: toPBTrafficSelector(tr.Destination),
		Target:      toPBTrafficTarget(tr.Target),
	}
}

func toPBTrafficSelector(ts *cce.TrafficSelector) *elapb.TrafficSelector {
	if ts == nil {
		return nil
	}

	return &elapb.TrafficSelector{
		Description: ts.Description,
		Macs:        toPBMACFilter(ts.MACs),
		Ip:          toPBIPFilter(ts.IP),
		Gtp:         toPBGTPFilter(ts.GTP),
	}
}

func toPBMACFilter(macf *cce.MACFilter) *elapb.MACFilter {
	if macf == nil {
		return nil
	}

	return &elapb.MACFilter{
		MacAddresses: macf.MACAddresses,
	}
}

func toPBIPFilter(ipf *cce.IPFilter) *elapb.IPFilter {
	if ipf == nil {
		return nil
	}

	return &elapb.IPFilter{
		Address:   ipf.Address,
		Mask:      uint32(ipf.Mask),
		BeginPort: uint32(ipf.BeginPort),
		EndPort:   uint32(ipf.EndPort),
		Protocol:  ipf.Protocol,
	}
}

func toPBGTPFilter(gtpf *cce.GTPFilter) *elapb.GTPFilter {
	if gtpf == nil {
		return nil
	}

	return &elapb.GTPFilter{
		Address: gtpf.Address,
		Mask:    uint32(gtpf.Mask),
		Imsis:   gtpf.IMSIs,
	}
}

func toPBTrafficTarget(target *cce.TrafficTarget) *elapb.TrafficTarget {
	if target == nil {
		return nil
	}

	return &elapb.TrafficTarget{
		Description: target.Description,
		Action:      toPBTargetAction(target.Action),
		Mac:         toPBMACModifier(target.MAC),
		Ip:          toPBIPModifier(target.IP),
	}
}

func toPBTargetAction(action string) elapb.TrafficTarget_TargetAction {
	switch action {
	case "accept":
		return elapb.TrafficTarget_ACCEPT
	case "reject":
		return elapb.TrafficTarget_REJECT
	case "drop":
		return elapb.TrafficTarget_DROP
	default:
		panic(fmt.Sprintf("invalid target action %s", action))
	}
}

func toPBMACModifier(macMod *cce.MACModifier) *elapb.MACModifier {
	if macMod == nil {
		return nil
	}

	return &elapb.MACModifier{
		MacAddress: macMod.MACAddress,
	}
}

func toPBIPModifier(ipMod *cce.IPModifier) *elapb.IPModifier {
	if ipMod == nil {
		return nil
	}

	return &elapb.IPModifier{
		Address: ipMod.Address,
		Port:    uint32(ipMod.Port),
	}
}

File: ui/controller/src/components/schema/TrafficPolicy.js
const IPFilter = {
  ip_filter: {
    title: "IP Filter",
    type: "object",
    properties: {
      address: {
        title: "IP Address",
        type: "string"
      },
      mask: {
        title: "Mask",
        type: "number",
        minimum: 0,
        maximum: 128,
      },
      begin_port: {
        title: "Begin Port",
        type: "number",
        minimum: 0,
        maximum: 65535,
      },
      end_port: {
        title: "End Port",
        type: "number",
        minimum: 0,
        maximum: 65535,
      },
      protocol: {
        title: "Protocol",
        type: "string",
        enum: [
          "all",
          "tcp",
          "udp",
          "icmp",
          "sctp"
        ]
      },
    },
  },
};

const MACFilter = {
  mac_filter: {
    title: "MAC Filter",
    type: "object",
    properties: {
      mac_addresses: {
        title: "MAC Addresses",
        type: "array",
        items: {
          title: "MAC Address",
          type: "string"
        }
      }
    }
  },
};

const GTPFilter = {
  gtp_filter: {
    title: "GTP Filter",
    type: "object",
    properties: {
      address: {
        title: "Address",
        type: "string"
      },
      mask: {
        title: "Mask",
        type: "number",
        minimum: 0,
        maximum: 128,
      },
      imsis: {
        title: "IMSIs",
        type: "array",
        items: {
          title: "IMSI",
          type: "number"
        }
      }
    }
  }
};

export default {
  schema: {
    type: "object",
    title: "Traffic Policy",
    required: [
      "name",
      "traffic_rules"
    ],
    properties: {
      id: {
        title: "ID",
        type: "string",
        readOnly: true
      },
      name: {
        title: "Name",
        type: "string",
        required: true
      },
      traffic_rules: {
        type: "array",
        title: "Traffic Rules",
        items: {
          type: "object",
          title: "Traffic Rule",
          properties: {
            description: {
              title: "Description",
              type: "string",
            },
            priority: {
              title: "Priority",
              type: "number",
              min: 0,
              max: 65535,
              required: true,
            },
            source: {
              title: "Source",
              type: "object",
              properties: {
                description: {
                  title: "Description",
                  type: "string"
                },
                ...MACFilter,
                ...IPFilter,
                ...GTPFilter,
              }
            },
            destination: {
              title: "Destination",
              type: "object",
              properties: {
                description: {
                  title: "Description",
                  type: "string"
                },
                ...MACFilter,
                ...IPFilter,
                ...GTPFilter,
              }
            },
            target: {
              title: "Target",
              type: "object",
              properties: {
                description: {
                  title: "Description",
                  type: "string",
                },
                action: {
                  title: "Action",
                  type: "string",
                  default: "accept",
                  enum: [
                    "accept",
                    "reject",
                    "drop"
                  ]
                },
                mac_modifier: {
                  title: "MAC Modifier",
                  type: "string"
                },
                ip_modifier: {
                  title: "IP Modifier",
                  type: "object",
                  properties: {
                    address: {
                      title: "IP Address",
                      type: "string"
                    },
                    port: {
                      title: "Port",
                      type: "number"
                    },
                  },
                },
              }
            }
          }
        }
      }
    }
  },
  form: [
    "*"
  ]
};
